testserv.c

```diff
@@ -87,12 +87,47 @@ mustdiallocal(int port)
     return fd;
 }
 
-#define SERVER() (progname = __func__, mustforksrv())
+static void
+exit_process(int signum)
+{
+    exit(0);
+}
+
+static void
+set_sig_handler()
+{
+    struct sigaction sa;
+
+    sa.sa_flags = 0;
+    int r = sigemptyset(&sa.sa_mask);
+    if (r == -1) {
+        twarn("sigemptyset()");
+        exit(111);
+    }
+
+    sa.sa_handler = exit_process;
+    r = sigaction(SIGTERM, &sa, 0);
+    if (r == -1) {
+        twarn("sigaction(SIGTERM)");
+        exit(111);
+    }
+}
+
+// Kill the child process with our chosen signal to give it a chance
+// to write gcov data to the filesystem before ct kills it with SIGKILL.
+static void
+kill_srv_atexit(void)
+{
+    kill(srvpid, SIGTERM);
+    waitpid(srvpid, 0, 0);
+}
+
+#define SERVER() (progname=__func__, mustforksrv(true))
+#define SERVER_NO_ATEXIT() (progname=__func__, mustforksrv(false))
```

Comment:
I wonder if it would be any better to test if srvpid is nonzero in kill_srv_atexit, so that we can install the atexit function unconditionally.

```c
static void
kill_srv_atexit(void)
{
    if (srvpid == 0) {
        return;
    }
    kill(srvpid, SIGTERM);
    waitpid(srvpid, 0, 0);
    srvpid = 0;
}
```

I think this logic would make it safe & okay for kill_srv_atexit to be registered two or more times.

(In that case it might also make sense to extract those last three lines into a helper function so it can be used in tests like cttestbinlogemptyexit.)

Just a thought. The `SERVER_NO_ATEXIT` approach is okay with me too.

Reply:
Very good observation!